### ISSUE TITLE
Fig 4 cumulative percentage of latent variables at a given resolution

### ISSUE THREAD

**Issue by betterze** [openai/vdvae#2]
Dear Adityaramesh,

Thank you for sharing with us this wonderful paper and implementation.

Would you mind telling me how to generate Fig 4 in the paper (Cumulative percentage of latent variables at a given resolution) using this implementation? I am new to VAE, sorry to ask silly questions.

Thank you for your help.

Best Wishes,

Alex

**Comment by rewonc** [comment]
Hi Alex,

This is pseudocode, so you might need to tweak a little:

```
n_batch = x.shape[0]
stochastic_layer_stats = vae.forward_get_latents(x)
latents = [stat['z'] for stat in stochastic_layer_stats]
K = 15
samples = vae.forward_samples_set_latents(n_batch=n_batch, latents=latents[:K])
```

This should sample images using only the first K=15 layers of latent variables from the ground truth, and sampling from the prior for the rest of the latents.

Feel free to tag me if you have other questions (so I get an alert about them).

Rewon

**Comment by betterze** [comment]
@rewonc Thx you for your guidance. The code works for me.  

Would you mind clarifying that what is the difference between 'activation' and 'latent'? Why the output of encoder ('activation') is not latent codes? Sorry, I new to VAE, would you mind providing the correspondence between the variable name in the code and the variable name in Fig 3? I believe it will be very helpful to understand the structure of this model.

And I try to use following code to generate random images, but the images are not very realistic. If I set the temperature 't' to be lower than 0.6, the images seem to be oversmooth. Is this the correct way to generate random images?

```
    tmp1=self.forward_uncond_samples(1,t=None)
    Image.fromarray(tmp1[0])
```
Thank you very much for your help.

**Comment by rewonc** [comment]
Activations are from the "bottom-up" path in the encoder. They are features
of the original image. The decoder and encoder will jointly generate latent
variables in "top-down" direction, using those encoder features. The Ladder
VAE paper discusses this in more detail:
https://arxiv.org/pdf/1602.02282.pdf.  You can think of latent variables as
being "noisy" versions of the image, whereas activations are noiseless
features of the image. The VAE reconstructs from the noisy versions, rather
than the noiseless features. (Reconstructing from noiseless features would
not require it to learn anything useful).

Yes, that's the correct way. t=0.8 or t=0.9 will probably give you the best
results. In the paper, we picked the best image at t=1 from a grid of 16
images. Also, make sure you're using the ema weights, rather than the
normal weights.

**Comment by betterze** [comment]
Thank you for your reply. I will read the paper.

**Comment by AranKomat** [comment]
@rewonc I'm runinng a small vd-vae on Imagenet32 with and without EMA, and I haven't seen much difference so far in terms of test ELBO. In what way does EMA help the training (e.g. as a stabilizer, a regularizer etc.)?

Btw your model is awesome! I think this will replace VQ-VAE for various models, which I'm trying to do in addition to some scaling analysis.

**Comment by rewonc** [comment]
Thanks for the feedback! EMA averages the model's weights, and typically
you will see a significant gain in log likelihood and sample quality toward
the end of training. You could alternatively anneal the learning rate
towards 0 to achieve a similar effect, but that can be harder to tune.
